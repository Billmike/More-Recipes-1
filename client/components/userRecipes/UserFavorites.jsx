import React from 'react';
import PropTypes from 'prop-types';

import UserFavoriteCard from './UserFavoriteCard';

const propTypes = {
  fetchUserFavorites: PropTypes.func.isRequired,
  currentProfileUserId: PropTypes.number.isRequired,
  userRecipes: PropTypes.shape({
    user: PropTypes.shape({
      userFavorites: PropTypes.arrayOf(PropTypes.shape())
    })
  }).isRequired,
};

/**
 * Class representing user favorite recipes
 *
 * @class UserFavorites
 *
 * @extends {React.Component}
 */
class UserFavorites extends React.Component {
  /**
   * Component did mount lifecycle method
   *
   * @returns {any} Fetches user favorite recipes
   *
   * @memberof UserFavorites
   */
  componentDidMount() {
    const { fetchUserFavorites, currentProfileUserId } = this.props;

    fetchUserFavorites(currentProfileUserId);
  }

  /**
   * Render method
   *
   * @returns {object} React element
   *
   * @memberof UserFavorites
   */
  render() {
    const { userRecipes: { userFavorites } } = this.props;

    return (
      <div className="row">
        {userFavorites &&
          userFavorites.length === 0
          ? <div className="center-align" >
            <h5>You have not favorited any recipe</h5>
            <i className="material-icons large">folder_open</i>
          </div>
          : userFavorites.map(favorite => (
            <UserFavoriteCard
              key={favorite.Recipe.id}
              recipe={favorite.Recipe}
              owner={favorite.Recipe.User}
              {...this.props}
            />
          ))}
      </div>
    );
  }
}

UserFavorites.propTypes = propTypes;

export default UserFavorites;
