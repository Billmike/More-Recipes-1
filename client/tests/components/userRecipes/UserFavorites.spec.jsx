import React from 'react';

import UserFavorites from '../../../components/userRecipes/UserFavorites';

describe('UserFavorites  component', () => {
  let spy;

  afterEach(() => {
    spy.mockClear();
  });
  it('should render component', () => {
    spy = jest.spyOn(UserFavorites.prototype, 'componentDidMount');
    const props = {
      fetchUserFavorites: jest.fn(),
      currentProfileUserId: 1,
      userRecipes: {
        isFetchingUserFavorites: false,
        userFavorites: [],
      },
    };
    const wrapper = shallow(<UserFavorites {...props} />);

    expect(wrapper).toBeDefined();
    expect(spy).toHaveBeenCalled();
    expect(wrapper.find('h5').text()).toEqual('No favorite recipes yet');
    expect(wrapper).toMatchSnapshot();
  });
  it('should show spinner', () => {
    const props = {
      fetchUserFavorites: jest.fn(),
      currentProfileUserId: 1,
      userRecipes: {
        isFetchingUserFavorites: true,
        userFavorites: [],
      },
    };
    const wrapper = shallow(<UserFavorites {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper.find('Spinner').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });
  it('should show user favorite recipes', () => {
    const props = {
      fetchUserFavorites: jest.fn(),
      currentProfileUserId: 1,
      userRecipes: {
        isFetchingUserFavorites: false,
        userFavorites: [{
          Recipe: {
            id: 2,
            title: 'Beans',
            category: 'Main',
            description: 'Tasty beans',
            preparationTime: 85,
            ingredients: 'Beans, onions, palm oil, salt',
            directions: 'Do this. Do that',
          },
        }],
      },
    };
    const wrapper = shallow(<UserFavorites {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper.find('UserFavoriteCard').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });
});
