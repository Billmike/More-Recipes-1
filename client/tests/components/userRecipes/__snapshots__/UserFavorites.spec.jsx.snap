// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserFavorites  component should render component 1`] = `
<div
  className="row"
>
  <div>
    <div
      className="center-align"
    >
      <h5>
        No favorite recipes yet
      </h5>
      <i
        className="material-icons large"
      >
        folder_open
      </i>
    </div>
  </div>
</div>
`;

exports[`UserFavorites  component should show spinner 1`] = `
<div
  className="row"
>
  <div
    className="center-align"
  >
    <Spinner
      size="big"
    />
  </div>
</div>
`;

exports[`UserFavorites  component should show user favorite recipes 1`] = `
<div
  className="row"
>
  <div>
    <UserFavoriteCard
      authenticatedUserId={null}
      currentProfileUserId={1}
      fetchUserFavorites={
        [MockFunction] {
          "calls": Array [
            Array [
              1,
            ],
          ],
        }
      }
      key="2"
      recipe={
        Object {
          "category": "Main",
          "description": "Tasty beans",
          "directions": "Do this. Do that",
          "id": 2,
          "ingredients": "Beans, onions, palm oil, salt",
          "preparationTime": 85,
          "title": "Beans",
        }
      }
      userRecipes={
        Object {
          "isFetchingUserFavorites": false,
          "userFavorites": Array [
            Object {
              "Recipe": Object {
                "category": "Main",
                "description": "Tasty beans",
                "directions": "Do this. Do that",
                "id": 2,
                "ingredients": "Beans, onions, palm oil, salt",
                "preparationTime": 85,
                "title": "Beans",
              },
            },
          ],
        }
      }
    />
  </div>
</div>
`;
